src/ucp/rndv/rndv_ppln.c

```diff
@@ -255,16 +256,14 @@ static ucs_status_t ucp_proto_rndv_ppln_progress(uct_pending_req_t *uct_req)
             return UCS_OK;
         }
 
-        /* Avoid sending message less than `frag_proto_min_length` */
-        next_frag = rpriv->frag_size;
-        msg_left  = req->send.state.dt_iter.length - req->send.state.dt_iter.offset;
-        if ((msg_left > rpriv->frag_size) &&
-            (msg_left < (rpriv->frag_size + rpriv->frag_proto_min_length))) {
-            next_frag = rpriv->frag_proto_min_length;
-        }
+        /* Avoid sending message less than `frag_proto_min_length` by
+           overlapping previous fragment */
+        req->send.state.dt_iter.offset = ucs_min(
```

Comment:
maybe encapsulate it in a new datatype_iter function

Reply:
yes, maybe add a variant of ucp_datatype_iter_next_slice that also accepts min_frag and returns overlap value (which is 0 if no overlap or not last frag)

Reply:
Done